src/frontend/js/modules/table-of-content.js

```diff
@@ -0,0 +1,217 @@
+/**
+ * Generate dynamic table of content
+ */
+export default class TableOfContent {
+  /**
+   * Initialize table of content
+   *
+   * @param {string} tagSelector - selector for tags to observe
+   * @param {string} tocWrapperSelector - selector for table of content wrapper
+   */
+  constructor({ tagSelector, tocWrapperSelector }) {
+    this.tagSelector = tagSelector || 'h2,h3,h4';
+    this.tocWrapperSelector = tocWrapperSelector;
+
+    this.init();
+  }
+
+  /**
+   * Initialize table of content
+   */
+  init() {
+    this.findTagsOnThePage();
+
+    /**
+     * Check if no tags found
+     */
+    if (this.tags.length === 0) {
+      console.info('Table of content is not needed');
+      return;
+    }
+
+    this.createTableOfContent();
+    this.addTableOfContent();
+    this.initIntersectionObserver();
+  }
+
+  /**
+   * Find all tags on the page
+   */
+  findTagsOnThePage() {
+    this.tags = Array.from(document.querySelectorAll(this.tagSelector));
+  }
+
+  /**
+   * Create table of content
+   *
+   * <section>
+   *   <header>On this page</header>
+   *   <ul>
+   *     <li><a href="#"></a></li>
+   *     ...
+   *   </ul>
+   * </section>
+   */
+  createTableOfContent() {
+    this.tocElement = document.createElement('section');
+    this.tocElement.classList.add('table-of-content__list');
```

Comment:
pls move all CSS classes to the constant, or import them directly from the CSS file using [CSS modules](https://github.com/css-modules/css-modules). Example: https://github.com/editor-js/footnotes-tune/blob/main/src/index.ts#L50